modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlValidator.java

```diff
@@ -524,15 +524,19 @@ private boolean isSystemFieldName(String alias) {
 
     /** {@inheritDoc} */
     @Override protected void inferUnknownTypes(RelDataType inferredType, SqlValidatorScope scope, SqlNode node) {
-        if (node instanceof SqlDynamicParam && inferredType.equals(unknownType)) {
-            if (parameters.length > ((SqlDynamicParam)node).getIndex()) {
-                Object param = parameters[((SqlDynamicParam)node).getIndex()];
+        if (node instanceof SqlDynamicParam ) {
+            if (inferredType.equals(unknownType)) {
+                if (parameters.length > ((SqlDynamicParam)node).getIndex()) {
+                    Object param = parameters[((SqlDynamicParam)node).getIndex()];
 
-                setValidatedNodeType(node, (param == null) ? typeFactory().createSqlType(SqlTypeName.NULL) :
-                    typeFactory().toSql(typeFactory().createType(param.getClass())));
+                    setValidatedNodeType(node, (param == null) ? typeFactory().createSqlType(SqlTypeName.NULL) :
```

Comment:
`typeFactory().createSqlType(SqlTypeName.NULL) ` might be cached.

Reply:
In `IgniteSqlValidator` for example. No need to create type every time for every null param.

Reply:
Planning is a resource consuming process, types caching during planning does not give any performance boost. I don't know why, but types are not cached in Calcite anywhere. On the other hand we cache it during execution (see ExpressionFactoryImpl#nullType), since it's more critical here.